<h1><center> 알고리즘으로 배우는 Python</center></h1>


## 목차

### 1. print, help, comment 처리

### 2. 사칙연산
- +, -, /, //, %, $*$, $**$ 

### 3. Python 의 기본 Data Type
- Integer
- float
- string
- bool

### 4. 파이썬의 변수 (variable), 표현식 (Expression), 문장 (Statement)
- naming 규칙
- data type 변환  
- 데이터 타입 변환 (Type Conversion)

### 5. 함수 (Function)
- 함수 정의 방법
- arguments, key word arguments
- name scope
- 내장함수
- 익명 함수 lambda
- 내장 sort 함수  

### 6. List 와 Tuple
- list 자료구조
- list indexing 방법
- list slicing
- tuple 자료구조
- tuple 정렬  
- zip  

### 7. for-loop
- 반복문 처리

### 8. 조건문
- 조건문 (if statement) 과 Boolean 자료형
- 비교연산자  
- 논리연산

### 9. range() 함수
- range 함수
- enumerate 함수

### 10. Dictionary 자료형
- dictionary 자료구조
- dictionary 생성 / 활용  
- dictionary 정렬


### 11. String (문자열)
- 스트링 (String) 자료형과 string 함수
- Escape Sequence  

### 12. String 의 여러가지 사용 방법
- 주요 method
- 문자열 formatting


### 13. Module
- 내장 module 불러오기
- 사용자 작성 module 불러오기

### 14. Class
- class 와 method 정의
- Object Oriented Programming
- class 변수, instance 변수
- 상속
- 다형성 구현


### 15. 집합 (set)
- 집합 자료구조


### 16. while 문



### 17. File 입출력
- file open/close
- file read/write
- IO error 처리
- file path 지정
- with open 문
- request 함수
- 예외처리 구문
- 특정 예외 처리
- 일반적 예외 처리
- assert

### 18. 정규식 (Regular Expression)


### 19. Functional Programming
- 재귀함수
- map, reduce, filter 함수  
- decorator  
- generator  

### 20. List Comprehension
- List Comprehension 작성 방법
- List Comprehension 내의 조건 check


### 21. 이진탐색(Binary Search) 알고리즘

### 22. 자주 사용하는 Python Library 소개

### 23. BeautifulSoup 으로 Webpage 읽어오기

### 25. Graph Data Structure

### 30. 멀티스레딩, 멀티프로세싱
- Multi-threading vs. Multi-processing  
- Process 의 주소 공간 (Address Space)  
- Shared Memory 를 이용한 프로세스 간 상태 공유  
- Multiprocessing Pool (Map Reduce)  
- Lock을 이용한 동기성(Synchronization) 유지

### 31. Argment Parse

### 33. Tkinter

### 50. unittest
- TestCase  
- Fixture  
- assertion

### 55. Factory Method


## Capstone Project

### 1. Memoization 과 Dynamic Programming

### 2. Simple Calculator 개발

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 1. Hello Python (print, help, comment)

### print(&nbsp;)   

  
### help( ) 혹은 help(명령어) 

In [1]:
print("hello python world")

hello python world


In [2]:
print("파이썬의 세계에 오신 것을 환영합니다.")

파이썬의 세계에 오신 것을 환영합니다.


In [3]:
print("하나", "둘", "셋")

하나 둘 셋


In [4]:
print("하나",",", "둘",",","셋")

하나 , 둘 , 셋


In [5]:
print("one " + "two " + "three")

one two three


In [6]:
print("Good morning !")
print("See you tomorrow !")

Good morning !
See you tomorrow !


In [7]:
print("Good morning !", end=""); 
print("See you tomorrow !")

Good morning !See you tomorrow !


### comment

In [8]:
# 여러분은 print 문을 이용하여 원하는 메시지를 출력할 수 있게 되었습니다. # 은 comment 입니다.
#print("comment 는 수행되지 않습니다")

### docstring

- 여러줄로 자유롭게 기술
- program 어느 곳이나 위치 할 수 있으나 가능한 program 시작 부분에 위치하는 것을 권장함
- program 의 목적 (what to do) 에 대하여 문서화

In [9]:
""" 
따옴표 세개가 시작과 끝 부분에 있므면 docstring 이므로 수행되지 않습니다. 
알고리즘으로 배우는 파이썬 과정은 살아 있는 파이썬을 배우는 과정입니다.
주로 프로그램 앞부분에서 프로그램의 목적에 대하여 기술하는 용도 입니다. 
"""
print("docstring 은 이렇게 여러줄을 표시할 수 있습니다.")

docstring 은 이렇게 여러줄을 표시할 수 있습니다.


In [10]:
print("문장 가운데 \n 이 들어 있으면 줄바꿈이 일어납니다.")

문장 가운데 
 이 들어 있으면 줄바꿈이 일어납니다.


In [11]:
print("문장 가운데 \\n 이 들어 있으면 줄바꿈이 일어납니다.")

문장 가운데 \n 이 들어 있으면 줄바꿈이 일어납니다.


In [12]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 2. 사칙 연산 

### 파이썬의 사칙 연산 규칙은 일반 수학 연산 규칙과 동일

$$30 - (10 + 20) * 3 + \frac{100}{2^2}$$

### 변수를 이용한 연산도 동일한 규칙을 따른다

    +, - : 더하기, 빼기

    /, // : 나누기 (소숫점 포함, 몫구하기)
    
    % : 나머지
    
    $*$ : 곱하기
    
    $**$ : 거듭제곱
    
    ( ) 안의 계산을 먼저한다


In [1]:
30 - (10 + 20) * 3 + 100 / 2 ** 2

-35.0

In [2]:
a = 30
b = 10
c = 20
d = 3
x = 100
k = 2

y = a - (b + c) * d + x / k ** 2

print(y)

-35.0


In [3]:
x = x  + 1
print(x)

101


### cascading assignment (계단식 할당)

In [4]:
z = y = x = 1 + 2 + 3
print(x, y, z)

6 6 6


In [5]:
print (a, b)

30 10


### value swapping

In [6]:
a , b = b, a
print(a, b)

10 30


In [7]:
2 ** 3

8

### exponential

In [8]:
2 ** 3 ** 4

2417851639229258349412352

In [9]:
(2 ** 3) ** 4

4096

In [10]:
3 ** 400

70550791086553325712464271575934796216507949612787315762871223209262085551582934156579298529447134158154952334825355911866929793071824566694145084454535257027960285323760313192443283334088001

In [11]:
3.0 ** 400

7.055079108655333e+190

In [12]:
time = 50
day = time / 24
day

2.0833333333333335

### 나눗셈

**floor division**

In [13]:
day = time // 24
day

2

**modulo (나머지)**

In [14]:
remainder = 30 % 24
remainder

6

### +=, -=, *=, /=, %= : augmented assignment

In [15]:
x = 1

In [16]:
x = x + 1
x

2

In [17]:
x += 1
x

3

In [18]:
y = 10

In [19]:
y = y - 1
y

9

In [20]:
y -= 1
y

8

In [21]:
z = 5

In [22]:
z = z * 2
z

10

In [23]:
z *= 2
z

20

### print 함수 내의 연산

In [24]:
integer1 = 11
integer2 = 5
print(integer1 * integer2)

55


In [25]:
print(integer1 // integer2)

2


In [26]:
print(integer1 % integer2)

1


In [27]:
print(integer1 / integer2)

2.2


영어권 국가에서는 일반적으로 쉼표가 천 단위 구분 기호로 사용되는 반면 다른 많은 국가에서는 마침표가 천 단위 구분 기호로 사용됩니다. 쉼표와 마침표가 파이썬에서 다른 것에 사용된다는 사실을 감안할 때 밑줄을 구분 기호로 사용하기로 결정했습니다. (from version 3.6)

In [2]:
100_000_000 + 100

100000100

In [3]:
print(int('1_000_000'))
print(float('1_000_00.0'))

1000000
100000.0


<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 3. 파이썬의 기본 데이터 타입 (Data Types)

### 1. integer (정수)

### 2. float (부동소숫점)

### 3. string (문자)

### 4. bool (불린)


### * 데이터 타입은 type(&nbsp;) 함수를 이용하여 알 수 있다.

In [1]:
7

7

In [2]:
type(7)

int

In [3]:
3.14

3.14

In [4]:
type(3.14)

float

In [5]:
+2.0

2.0

In [6]:
type(+2.0)

float

In [7]:
type(2)

int

$1.23 x 10^3$

In [8]:
1.23e3

1230.0

In [9]:
type(1.23e3)

float

$1.0 \times 10^0$

In [10]:
1e0

1.0

In [11]:
type(1e0)

float

In [12]:
type('A')

str

In [13]:
"문자형은 string 데이터 타입이다."

'문자형은 string 데이터 타입이다.'

In [14]:
type("문자형은 string 데이터 타입이다.")

str

In [15]:
type('따옴표 하나도 string 이다')

str

In [16]:
'1.23'

'1.23'

In [17]:
type('1.23')

str

### print 함수 내의 number 는 string 으로 변환 후 출력

In [18]:
print(1.23, "3.14")

1.23 3.14


In [19]:
type(True)

bool

In [20]:
type(False)

bool

In [21]:
2 == 2

True

In [22]:
2 == 3

False

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 4. 파이썬의 변수 (variable), 표현식 (Expression), 문장 (Statement)

## 변수 (variable)

- 사전에 정해진 type 이 없고 보관되는 data type 에 따라 dynamic 하게 정해진다.  


-  변수의 이름은 문자와 숫자, \_ (underscore) 로 구성되며, 문자 (a\~z, A\~Z) 혹은 \_ (underscore) 로만 시작 가능하다.  


-  = 를 이용하여 변수에 값을 할당 한다  


-  Python 이 사용하는 33 개의 keyword 는 변수명으로 사용하면 안된다.

    False, None, True, and, as, assert, break, class, continue, def, del, elif, else, except,finally, for, from, global, if, import, in, is, 
    lambda, nonlocal, not, or, pass, raise, return, try, while, with, yield
    
## 표현식 (Expression)

- 변수와 operation 의 조합  

- 예) 17, x, x + 17

## 문장 (Statement)

- Python interpreter 가 수행할 수 있는 code 의 단위 

- 예) print(17), print(x), x = 17

In [1]:
n = 5
print(n)

5


In [2]:
x = 3.14
print(x)

3.14


In [3]:
a = "Python's data type is powerful and easy."
print(a)

Python's data type is powerful and easy.


In [4]:
a = 100
x = 55

In [5]:
print("n =", n, ", x =", x, ", a = ", a)

n = 5 , x = 55 , a =  100


### 변수명

In [6]:
_one2three = 123

In [7]:
one_2_three = 123

In [8]:
one.2.three = 123

SyntaxError: invalid syntax (<ipython-input-8-bb2f283060e8>, line 1)

In [9]:
1_two_three = 123

SyntaxError: invalid token (<ipython-input-9-82e7f50b07f1>, line 1)

### type() 함수는 Python 의 자료형을 반환

In [10]:
my_integer = 50
print(my_integer, type(my_integer))

50 <class 'int'>


In [11]:
my_float = 50.0
print(my_float, type(my_float))

50.0 <class 'float'>


In [12]:
my_float = float(50)
print(my_float, type(my_float))

50.0 <class 'float'>


In [13]:
my_integer = int(3.14)
print(my_integer, type(my_integer))

3 <class 'int'>


In [14]:
my_string = str(3.14)
print(my_string, type(my_string))

3.14 <class 'str'>


In [15]:
my_bool = True
print(my_bool, type(my_bool))

True <class 'bool'>


### 데이터 타입 변환 (Type Conversion)

In [16]:
int("123")

123

In [17]:
float("23.5")

23.5

In [18]:
float("inf")

inf

In [19]:
str(123)

'123'

In [20]:
int(True)

1

### eval() 함수 : expression 을 evaluate 하고 result 를 return

In [21]:
eval("1 + 2 + 3")

6

In [22]:
a = 10
b = 20
eval("a + b")

30

# 연습문제

1) 변수 x 의 값이 10 이고 변수 y 의 값이 20 인 경우 다음 문장의 수행 후 x, y 의 값은 ?

    x = y
    y = x
    print(x, y)

2) x 값이 3 이라고 할때 x ** 2 + x * x 의 값은 ?  

3) 2 + 4 % 5 의 값은 ?

4) 다음 두 연산식의 결과는 ?

    5.5 - 11 / 2

    5.5 - 11 // 2
    
    5.5 - 11 % 2

5) % 를 이용한 홀수, 짝수 구분

6) 변수값의 할당

    hello = "안녕"

    world = "하세요"

    print(hello, world) 의 결과는 ?

7) 변수값의 할당

    x = 10
    y = 20

    a, b = x + 1, y + 1

    print(a,b) 의 결과는 ?

8) 원의 넓이 구하기 : 반지름이 10 인 원의 넓이를 $S = \pi r ^ 2 $ 공식을 사용하여 구한다. ($\pi = 3.14$)

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 5. 함수(function)

## 5-1. 함수의 정의 및 호출

**함수는 프로그램의 기능을 작은 task 단위로 나누어 재사용 가능한 독립적 형태로 나눈 것**

**함수는 결과값(return value)을 반환하는 함수와 결과값을 반환하지 않는 함수로 구분된다.**

    - 결과값을 반환하지 않는 파이썬 함수의 예 : print(&nbsp;)
    - 결과값을 반환하는 파이썬 함수의 예 : type(&nbsp;)

### 함수의 작성 방법:

**1. def keyword 로 시작한다.**
**2. 함수의 이름은 변수의 이름과 작성 규칙이 동일하다.**

        문자와 숫자, _ (underscore) 로 구성되며, 문자 (a~z, A~Z) 혹은 _ (underscore) 로만 시작 가능.  
        
**3. 함수 이름 다음에 ( ) 로 둘러싸여진 인수 (parameter, argument) 전달 가능**

**4. 들여쓰기 (indentation) 에 의해 함수의 body 임을 표시한다.**

**5. return keyword 를 이용하여 결과값을 반환한다. 반환할 값이 없으면 return 은 생략 가능하다.**
```
    def function_name (*args, **args):
        function_body  
        return (value)
```    
**6. 함수의 반환값은 변수에 저장할 수 있다.**

**7. 여러개의 값을 한번에 반환할 수 있다.** 

**8. Python 이 기본적으로 제공하는 함수를 내장함수 (built-in function) 이라고 한다.**

    print(), sum(), abs(), help(), str(), round(), type() 등 약 70 가지
    
### \*args, \**kwargs

- \*args 로 여러개의 parameter 를 한번에 받아들임  
- \**kwargs 로 여러개의 keyword argument 를 한번에 받아들임

In [1]:
def hello_world():
    print("hello world")
    
hello_world()

hello world


In [2]:
def return_result():
    a = 10
    b = 20
    result = a + b
    return result

x = return_result()
print(x)

30


In [3]:
def param_func(x, y):
    result = x / y
    return result

a = 10
b = 20
x = param_func(a, b)
print(x)

0.5


In [4]:
def circle(pi, radius):
    print("pi = ", pi)
    print("반지름 = ", radius)
    return pi * radius ** 2

In [5]:
circle(3.14, 10)

pi =  3.14
반지름 =  10


314.0

In [6]:
r1, r2 = param_func(a, b) * 20, circle(3.14, a) / 10
print(r1, r2)

pi =  3.14
반지름 =  10
10.0 31.4


In [7]:
def plus_and_multi(m, n):
    return m + n, m * n

x, y = plus_and_multi(5, 10)
print(x, y)

15 50


In [8]:
def calculate_volume(length, width, depth):
    return length * width * depth

calculate_volume(10, 20, 30)

6000

In [1]:
def sum_values(*args):
    print(args)

In [2]:
sum_values(1,2,3)

(1, 2, 3)


In [3]:
sum_values(1,2,3,4,5)

(1, 2, 3, 4, 5)


In [12]:
x = print("Hello World")

Hello World


In [13]:
print(x)

None


In [14]:
a = type(10)

In [15]:
print(a)

<class 'int'>


## 5-2. Scope of Variable (변수 영역)

### variable 에는 global variable (전역변수) 과 local variable (지역변수) 의 두가지 종류가 있음

### global variable - 함수의 바깥에 정의하고 프로그램 전체에서 접근 가능
### local variable - 함수의 안쪽에 정의하고 함수 내에서만 접근 가능
### gobal variable 과 local variable 이 같은 이름을 가질 경우는 local variable 이 우선함

### 함수의 영역(Scope of Function) 도 변수의 영역과 동일함.

In [17]:
a = 1
b = 2

def func():
    c = 3
    d = 4
    print("global variable :", a, b)
    print("local  variable :", c, d)

In [18]:
func()

global variable : 1 2
local  variable : 3 4


In [19]:
print(c)

NameError: name 'c' is not defined

In [20]:
price = 100

def nego(price):
    price = price - 10
    print("negotiated price =", price)
    
nego(price)
print(price)

negotiated price = 90
100


### 전역변수를 함수 내에서 update

- not recommendable  
- global keyword 사용

In [2]:
a = 15
  
# global value 변경
def change(): 
  
    global a
    
    a = a + 5 
    print(a) 

change() 

20


In [3]:
def outer(x):
    
    def inner(y):
        return y ** 2 + 2
    
    z = inner(x) + 2
    
    return z

print(outer(4))

20


In [4]:
inner(4)

NameError: name 'inner' is not defined

In [5]:
def outer(x):
    def inner(y):
        return y ** 2 + 2
    return inner(x) + 2

outer(4)

20

In [6]:
def square(length):
    print("정사각형의 한변 = ", length)
    return length ** 2

def outer(x):
    def inner(y):
        return y ** 2
    return inner(x) + square(x)

outer(4)

정사각형의 한변 =  4


32

In [27]:
def rectangle(L1, L2):
    return L1 * L2

print(rectangle(10, 20))
print(rectangle(L1=10, L2=20))

200
200


## 5-3. \*args, \**kwarg

In [28]:
def names(*args):
    print(args)
    
names('오', '영제', '남성')

('오', '영제', '남성')


In [29]:
def names(**kwargs):
    print(kwargs)

names(first='오', second='영제', third='남성')

{'first': '오', 'second': '영제', 'third': '남성'}


In [30]:
def names(*args, **kwargs):
    print(args)
    print(kwargs)
    
names('오', '영제', gender='남성')

('오', '영제')
{'gender': '남성'}


## 5-4. Python 의 내장함수 (Built-In Function)

#### Python 에 내장되어서 명령어처럼 사용할 수 있는 함수(function)들을 내장함수 (built-in function) 이라고 한다.

### print 함수

In [31]:
print(abs(-3))

3


In [32]:
print(bool(0))
print(bool(1))

False
True


### input 함수

In [36]:
year = input('태어난해를 입력하시오')

태어난해를 입력하시오2000


In [37]:
year

'2000'

### eval 함수

In [38]:
eval('10 * 5')

50

### 형변환 함수

In [39]:
float("12")

12.0

In [40]:
int('123')

123

In [41]:
lst = list((1,2,3))

### 기타

In [42]:
len(lst)

3

In [43]:
max(lst)

3

In [44]:
min(lst)

1

In [45]:
test_file = open("./test_text.txt")

In [46]:
test_file.read()

'This is a python open built-in function test file'

## 5-5. Python 의 익명 함수 lambda

- Python 의 함수 생성 방법은 def 와 lambda 두 가지가 있다.
- lambda 는 익명 (anonymous) 함수 이다.
- 한번 사용할 간단한 함수인 경우 사용

In [47]:
def sum(x, y):
    return x + y

In [48]:
sum(1, 2)

3

In [49]:
(lambda x, y: x+y)(1,2)

3

### lambda 를 변수에 assign

In [50]:
g = lambda x: x**2

print(g(8))

64


In [51]:
f = lambda x, y: x + y

print(f(4, 5))

9


## typing— 유형 힌트 지원 

- 버전 3.5의 새로운 기능.  


- Python 런타임은 함수 및 변수 유형 주석을 적용하지 않습니다. 유형 검사기, IDE, linter 등과 같은 3rd party 도구에서 사용할 수 있습니다.

In [1]:
def greeting(name: str) -> str:
    return 'Hello ' + name

In [5]:
greeting(123)

TypeError: can only concatenate str (not "int") to str

## 연습문제

1) 선형방정식 (linear equation) y = mx + b 를 함수로 작성한다. 

   이때, 기본값(default value)은 m = 1, b = 0 로 한다.

2) 다음 함수가 수행된 후 print 되는 값은 ?

    def f(x):
        return x + 1, x * x

    x, y = f(3)
    print(x, y)

3) 다음 code 가 수행된 이후 z 의 값은 ?

    def f1(x, y):
        return (x + 1) / (y - 1)

    z = f1(2, 2)
    print(z)
    
    def f1(x, y=2):
        return (x + 1) / (y - 1)

    z = f1(1)
    print(z)

4) 섭씨 온도를 화씨 온도로 변환하는 함수를 작성한다. 변환 공식은 다음과 같다.

$$ T_f = \frac{9}{5} T_c +32$$

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 6. 리스트(List) 와 Tuple

## 6.1 List 구조

- 일상생활에서 흔히 볼 수 있는 자료구조 (Data Structure)

        예) 할일 목록 (to-do list), 식료품 구입 리스트, 초대장 리스트, 소원목록 (wish list), 국가 목록 등.

-  List 는 top-10 list 처럼 순서가 있을 수도 있고, 친구 목록처럼 순서가 없을 수도 있음.


- List 의 항목들을 요소 (element) 라고 한다.


-  Python 문법에서 list 는 대괄호 (bracket) 로 표시하고 element 는 콤마(,) 로 구분하여 대괄호안에 위치함.


-  List 의 element 는 지금까지 배운 Python 의 모든 자료형과 함수가 사용될 수 있음.

In [12]:
x = ["apple", 'orange', "옥수수", '밀']

In [13]:
print(x)

['apple', 'orange', '옥수수', '밀']


In [14]:
y = [1, 'two', 6/2]

In [15]:
print(y)

[1, 'two', 3.0]


In [16]:
def f(n):
    return [n, n * 2, n * 3, n * 4]

In [17]:
print(f(3))

[3, 6, 9, 12]


In [18]:
y = f(4)

In [19]:
type(y)

list

In [20]:
f('hi')

['hi', 'hihi', 'hihihi', 'hihihihi']

In [21]:
a = 5
b = 3.5
x = [a, a+b, a-b]

In [22]:
print(x)

[5, 8.5, 1.5]


### list 의 주요 method

- len( ) : list 의 길이
- sort( ) : list sort
- reverse( ) : list 를 역순으로 sort
- append( ) : list 끝에 new item 추가
- extend( )  : 다른 list 추가
- pop( )  : list 의 마지막 item 제거 및 반환
- pop(i) : i 번째 item 제거 및 반환
- del alist[i] : i 번째  item delete
- count( ) : list element 의 갯수 
- index( ) : list element 의 index

### 여러개의 list 를 결합

In [23]:
c = ['z', 'y', 'w'] + ['a', 'c', 'b']
c

['z', 'y', 'w', 'a', 'c', 'b']

In [24]:
c.pop(2)

'w'

In [25]:
del c[2]

In [26]:
c

['z', 'y', 'c', 'b']

### list 의 원소(element) 갯수

In [27]:
len(c)

4

### list 원소(element) 추가 / 확장

In [28]:
d = []

In [29]:
print(d)

[]


In [30]:
len(d)

0

In [31]:
d.append('first')

In [32]:
d

['first']

In [33]:
d.extend(c)

In [34]:
d

['first', 'z', 'y', 'c', 'b']

In [35]:
e = d + c
print(e)

['first', 'z', 'y', 'c', 'b', 'z', 'y', 'c', 'b']


### list 내의 list

In [36]:
d.append(['second, third'])
d

['first', 'z', 'y', 'c', 'b', ['second, third']]

In [37]:
animals = [['dog', 'Ruby', 7], 
           ['cat', 'Nero', 3], 
           ['dog', 'John', 5], 
           ['bird', 'Seagal', 2]]
print(animals)

[['dog', 'Ruby', 7], ['cat', 'Nero', 3], ['dog', 'John', 5], ['bird', 'Seagal', 2]]


### list element 의 변수 assign

In [38]:
[x, y, z] = 1, 2, 3
print(x, y, z)

1 2 3


In [39]:
r, s, t = [1, 2, 3]
print(r, s, t)

1 2 3


In [40]:
u = [1, 2, 3]
j, k, l = u
print(j, k, l)

1 2 3


In [41]:
x = 1
y = 2
xlist = [x, y, x + y]
x = 0
y = 0

In [42]:
xlist

[1, 2, 3]

### list methods

In [43]:
print(c)

['z', 'y', 'c', 'b']


In [44]:
c.sort()

In [45]:
print(c)

['b', 'c', 'y', 'z']


In [46]:
c.sort(reverse=True)
c

['z', 'y', 'c', 'b']

In [47]:
sum(xlist)

6

In [48]:
max(xlist)

3

In [49]:
min(xlist)

1

In [50]:
tuple(xlist)

(1, 2, 3)

In [51]:
del(xlist[0])
xlist

[2, 3]

In [52]:
alist = [1,2,3,3,4,4,4,5]
alist.count(3)

2

In [53]:
alist.index(3)

2

### list 정렬

- list.sort()  
- sorted()

- Python 의 내장 sort 함수는 merge sort 와 insertion sort 를 결합한 time sort 알고리즘

In [1]:
a = [1, 10, 5, 7, 6]
a.sort()

print(a)

[1, 5, 6, 7, 10]


In [2]:
sorted([5, 2, 3, 1, 4])

[1, 2, 3, 4, 5]

In [3]:
sorted("This is a test string from Andrew".split(),  key=str.lower)

['a', 'Andrew', 'from', 'is', 'string', 'test', 'This']

## 6-2. List 의 indexing

### list 를 구성하는 element 들은 위치에 따라 index 로 접근할 수 있다.
### index 는 대괄호(bracket)안에 정수(integer) 로 표시
### Python 의 index 는 0 부터 시작하는 양수 혹은 -1 부터 시작하는 음수로 표시할 수 있다.
        0   1     2     3     4    5 
    +---+---+---+---+---+---+
    |  H |  e  |  l  |   l |  o |  ! |
    +---+---+---+---+---+---+
      -6    -5    -4   -3   -2   -1

In [54]:
hello = ['H', 'e', 'l', 'l', 'o', '!']
print(len(hello))

6


In [55]:
print(hello)

['H', 'e', 'l', 'l', 'o', '!']


In [56]:
hello[0]

'H'

In [57]:
hello[5]

'!'

In [58]:
hello[len(hello) - 1]

'!'

In [59]:
x = [1,2,3] + ['a','b','c']
print(x)

[1, 2, 3, 'a', 'b', 'c']


In [60]:
x[0]

1

In [61]:
x[len(x)-1]

'c'

In [62]:
x[-1]

'c'

In [63]:
x[3] = 4

In [64]:
print(x)

[1, 2, 3, 4, 'b', 'c']


In [65]:
x[4] = 5
x

[1, 2, 3, 4, 5, 'c']

In [66]:
x[0] = 'first'
x[-1] = 'last'
x

['first', 2, 3, 4, 5, 'last']

In [67]:
x[-2]

5

In [68]:
x[len(x) - 1] = 'end'
x

['first', 2, 3, 4, 5, 'end']

In [69]:
x[-6]

'first'

In [70]:
animals = [['dog', 'Ruby', 7], 
           ['cat', 'Nero', 3], 
           ['dog', 'John', 5], 
           ['bird', 'Seagal', 2]]

In [71]:
animals[2]

['dog', 'John', 5]

In [72]:
animals[2][2]

5

In [73]:
kind, name, age = animals[-1]

In [74]:
print(kind, name, age)

bird Seagal 2


## 6-3. 인덱스 (Index) 를 이용한 list 의 element 자르기 (slicing)


### list 의 일부분을 범위 지정할 수 있다.
    list[start : end]   => start ~ end-1
    list[start:]      => start ~ 끝까지
    list[:end]        => 처음 ~ end-1
    list[:]           => list 전체
    
### 증가분 (increment) 를 지정하면 더욱 다양한 slicing 이 가능하다.
    list[start : end : increment]
    
### increment 가 음수 (-) 이면 뒤에서부터 거꾸로 처리한다.

In [75]:
x = [1, 2, 3, 4, 5, 6]

In [76]:
y = x[0:2]
print(y)

[1, 2]


In [77]:
x[2:4]

[3, 4]

In [78]:
x[2:]

[3, 4, 5, 6]

In [79]:
x[:6]

[1, 2, 3, 4, 5, 6]

In [80]:
x[:]

[1, 2, 3, 4, 5, 6]

In [81]:
x[-3:]

[4, 5, 6]

In [82]:
x[2:-1]

[3, 4, 5]

In [83]:
x[-6:3]

[1, 2, 3]

In [84]:
s = [1,2,3,4,5,6,7,8,9,10]
python = ['p','y','t','h','o','n']

In [85]:
s[:: 2]

[1, 3, 5, 7, 9]

In [86]:
python[::2]

['p', 't', 'o']

In [87]:
s[5::2]

[6, 8, 10]

In [88]:
python[5::2]

['n']

In [89]:
s[5:-1:2]

[6, 8]

In [90]:
s[::-1]

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [91]:
s

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [92]:
s[7:1:-1]

[8, 7, 6, 5, 4, 3]

In [93]:
s[-1:-8:-1]

[10, 9, 8, 7, 6, 5, 4]

In [94]:
t = ['a','b','c']

In [95]:
t[1:100]

['b', 'c']

In [96]:
t[-2:-1]

['b']

In [97]:
t[-100:100]

['a', 'b', 'c']

In [98]:
t[-100]

IndexError: list index out of range

In [99]:
t[5]

IndexError: list index out of range

In [100]:
animals = [['dog', 'Ruby', 7], 
           ['cat', 'Nero', 3], 
           ['dog', 'John', 5], 
           ['bird', 'Seagal', 2]]

In [101]:
animals[1][0]

'cat'

In [102]:
animals[-1][2]

2

In [103]:
animals_tuple = (('dog', 'Ruby', 7), 
                 ('cat', 'Nero', 3), 
                 ('dog', 'John', 5), 
                 ('bird', 'Seagal', 2))

In [104]:
animals_tuple[1][0]

'cat'

In [105]:
animals_tuple[-1][2]

2

## 6-4. 튜플 (Tuple) 

### 튜플은 element 값을 변경할 수 없는 list 이다.

- Mutable : 값을 변경 가능 (list, variable)

- Immutable : 값을 변경 불가능 (tuple)  


- **변경할 수 없으므로 list 보다 memory 사용이 효율적**

### Tuple 은 괄호 (  ) 로 element 들을 감싸준다.

### element 를 변경할 수 없는 것 이외의 operation 은 list 와 거의 유사함.

### Tuple 과 List 비교

In [106]:
xlist = [1, 2, 3, 4]
ytuple = (1, 2, 3, 4)

In [107]:
xlist[1:4]

[2, 3, 4]

In [108]:
ytuple[1:4]

(2, 3, 4)

In [109]:
ytuple[-1]

4

In [110]:
xlist.append(6)
x

[1, 2, 3, 4, 5, 6]

In [111]:
ytuple.append(6)

AttributeError: 'tuple' object has no attribute 'append'

In [112]:
xlist[3] = '4th'
xlist

[1, 2, 3, '4th', 6]

In [113]:
ytuple[3] = '4th'

TypeError: 'tuple' object does not support item assignment

In [114]:
(x, y) = (1, 'two')

print(x, y)

1 two


In [115]:
(x, y) = [1, 'two']

print(x, y)

1 two


In [116]:
[x, y] = [1, 'two']

print(x, y)

1 two


### 원소가 한개 (single element) 인 Tuple
- comma(,) 로 tuple 임을 표시

In [117]:
t1 = ('a',)
type(t1)

tuple

### 여러개의 Tuple 을 하나로 merge

- 기존의 tuple 을 변경하는 것이 아니고 새로운 tuple 을 생성

In [118]:
z = ('a','b')

In [119]:
new_tuple = ytuple + z
new_tuple

(1, 2, 3, 4, 'a', 'b')

### Tuple 생성

In [120]:
t = 1, "two", 3+4

In [121]:
t

(1, 'two', 7)

In [122]:
type(t)

tuple

In [123]:
j, k, l = u
print(j, k, l)

1 2 3


### tuple method

In [124]:
t = tuple('abcde')
t

('a', 'b', 'c', 'd', 'e')

In [125]:
t = tuple(['a', 'b', 'c', 'd', 'e'])
t

('a', 'b', 'c', 'd', 'e')

### Tuple assignment

In [126]:
my_tuple = 'cat', 'dog', 'apple'
type(my_tuple)

tuple

In [127]:
cc, dd, aa = my_tuple
print(cc, dd, aa)

cat dog apple


In [128]:
a, b = 1, 2, 3

ValueError: too many values to unpack (expected 2)

### Tuple indexing

In [129]:
print(my_tuple[-1])
print(my_tuple[-2])

apple
dog


In [130]:
my_tuple[-1] = 'banana'

TypeError: 'tuple' object does not support item assignment

### Tuple 정렬 (sort)

In [131]:
new_tuple = my_tuple + ('banaba',)

In [132]:
new_tuple

('cat', 'dog', 'apple', 'banaba')

In [133]:
new_tuple.sort()

AttributeError: 'tuple' object has no attribute 'sort'

In [134]:
sorted(new_tuple)

['apple', 'banaba', 'cat', 'dog']

In [4]:
student_tuples = [
    ('john', 'A', 15),
    ('jane', 'B', 12),
    ('dave', 'B', 10),
]

sorted(student_tuples, key=lambda student: student[2])   # sort by age

[('dave', 'B', 10), ('jane', 'B', 12), ('john', 'A', 15)]

In [5]:
sorted(student_tuples, key=lambda student: student[2], reverse=True)  

[('john', 'A', 15), ('jane', 'B', 12), ('dave', 'B', 10)]

### Tuple 의 크기 비교 

- 왼쪽 element 부터 비교

In [135]:
(0, 8, 9) < (1, 2, 3)

True

In [136]:
(0, 8, 9) < (0, 2, 3)

False

In [137]:
('a', 'b') < ('abc', 'ab')

True

## zip

- zip(*iterable)은 동일한 개수로 이루어진 자료형을 묶어 주는 역할을 하는 함수

In [138]:
list(zip([1, 2, 3], [4, 5, 6]))

[(1, 4), (2, 5), (3, 6)]

In [139]:
list(zip([1, 2, 3], [4, 5, 6], [7, 8, 9]))

[(1, 4, 7), (2, 5, 8), (3, 6, 9)]

In [140]:
list(zip("abc", "def"))

[('a', 'd'), ('b', 'e'), ('c', 'f')]

In [141]:
fruits = ['apple', 'orange', 'pear', 'grape']
prices = [100, 200, 150, 50]

for f, p in zip(fruits, prices):
    print(f, p)

apple 100
orange 200
pear 150
grape 50


## List / Tuple 연습문제

1) 다음 프로그램의 결과값은 ?

    xlist = []
    xlist.append('Good')
    xlist.append('Morning')
    print(xlist)
    
    xlist.append([3, 4])
    print(xlist)

2) 다음 list 의 element 를 오름차순으로 정렬 (ascending sort) 한다. 또한, 내림차순 (descending order)으로 정렬한다. sort(), sort(reverse=True) 함수를 사용한다.
    
    xlist = [2, 1, 3, 5, 4]

3) 두개의 list element 들을 짝을 지워 출력  

stocks = ['삼성전자', '대한항공', 'google', 'apple']  
close = [40000, 2000, 50000, 100000]

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 7. for-loop (반복문)

#### for 문을 이용하여 반복 처리를 할 수 있다.

#### Python 에서는 for-loop 을 이용하여 list 내의 element 들을 순차적으로 처리할 수 있다.
#### list 와 같이 한번에 한개씩의 element 를 반환하는 자료형태를 iterable (반복형태) 이라고 부른다.
#### Python 의 iterable 은 list, tuple, dictionary 이다.
#### Python 의  for loop 의 문법은 다음과 같다.
```
for 변수 in 리스트(또는 튜플, 문자열):  
    수행할 문장1  
    수행할 문장2
```

### list 와 for-loop

In [1]:
fruits = ["apple", "orange", "banana", "mango", "pear"]

In [2]:
for fruit in fruits:
    print("I love", fruit)

I love apple
I love orange
I love banana
I love mango
I love pear


In [3]:
count = 0
for _ in fruits:
    count += 1
print(count)

5


In [4]:
x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
sum = 0
for i in x:
    print(i)
    sum = sum + i
print(sum)

1
2
3
4
5
6
7
8
9
10
55


In [5]:
for i in x:
    print(i, end=" ")

1 2 3 4 5 6 7 8 9 10 

In [6]:
for i in x:
    print(i, end="")

12345678910

In [7]:
list_items = ['1st', '2nd', '3rd', '4th', '5th', '6th']
index = [0,1,2,3,4,5]
for i in index:
    print(list_items[i])

1st
2nd
3rd
4th
5th
6th


## For-Loop 연습문제

1) 다음 문장 수행 후의 output 은 ? 

    xlist = [1, [1, 2], [1, 2, 3]]
    print(xlist[1][1] + 1)

2) xlist = [1, 2, 3, 4] 를 한줄에 출력하고 각 element 의 사이는 '/' 로 구분하라.

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 8. 조건문 

## 8-1. 조건문 (if statement) 과 Boolean 자료형

### boolean 자료형 :  True, False, bool( )

### 조건이 참(True) 인 경우만 명령 수행

### 문법 :
```    
    if <test 문>:
        <body1>
    elif <test 문>:
        <body2>
    else:
        <body3>
```        
### Python 의 False 와 True 

    False : empty string(""), None, [], (), 0, False
    True  : 그 외의 모든 value

In [1]:
if True:
    print("Test passed")

Test passed


In [2]:
if False:
    print("Test failed")

In [3]:
i = 3
if i == 1:
    print('i = 1')
elif i == 2:
    print('i = 2')
elif i == 3:
    print('i = 3')
else:
    print('i > 3')

i = 3


In [4]:
i = 10
if i % 2 == 0:
    if i % 3 == 0:
        print('i 는 2 와 3 의 배수')
    if i % 5 == 0:
        print('i 는 2 와 5 의 배수')
    else:
        print('i 는 2 의 배수이지만 3 이나 5 의 배수는 아님')
else:
    print('i 는 홀수임')

i 는 2 와 5 의 배수


In [5]:
bb = True
type(bb)

bool

In [6]:
if bb:
    print("True")

True


In [7]:
x = 0; y = 0.0; z = 0 + 0j
bool(x), bool(y), bool(z)

(False, False, False)

In [8]:
x = -1; y = 1.e-10; z = 0 + 1j
bool(x), bool(y), bool(z)

(True, True, True)

In [9]:
u = False; v = 0; w = ''; x = []; y = (); z = None        # None : No Value
bool(u), bool(v), bool(w), bool(x), bool(y), bool(z)

(False, False, False, False, False, False)

In [10]:
u = True; v = 1; w = ' '; x = [0]; y = (0,0); z = "0"
bool(u), bool(v), bool(w), bool(x), bool(y), bool(z)

(True, True, True, True, True, True)

In [11]:
items = ["", None, [], 0, False, " ", [""], (), 1.5, (0, 0)]
true  = [] 
false = []
for item in items:
    if item:
        true.append(item)
    else:
        false.append(item)
print("true  =", true)
print("false =", false)

true  = [' ', [''], 1.5, (0, 0)]
false = ['', None, [], 0, False, ()]


## 8-2. 비교 (Comparison)

### 비교연산자 8 가지 :

    x <  y 
    x <= y
    x == y   (두 변수의 value 가 같은지 비교) 
    x >= y
    x >  y
    x != y
    x in y
    x is y     (두 object 가 동일한지 비교)

In [1]:
a = [1, 2, 3]
b = [1, 2, 3]
print(a == b)
print(a is b)

True
False


In [13]:
a = 2
if a == 2:
    print("True")

True


In [14]:
if a == 3:
    print("True")
else:
    print("False")

False


In [15]:
7 == 7.0

True

In [16]:
x = 0.5
1 == 2 * x

True

In [17]:
7 != "7"

True

In [18]:
3 >= 2

True

In [19]:
5 < 7

True

In [20]:
if 3 * 5 == 15:
    print("3 * 5 = 15")

3 * 5 = 15


In [22]:
xlist = [1,2,3,4,5,6,7]

for element in xlist:
    if element >= 5:
        print(element)

5
6
7


In [23]:
2 in xlist

True

In [24]:
xlist[1] is 2

True

In [25]:
animals = [['dog', 'Ruby', 7], 
                 ['cat', 'Nero', 3], 
                 ['dog', 'John', 5], 
                 ['bird', 'Seagal', 2]]

In [26]:
for kind, name, age in animals:
    if age > 3:
        print(kind, name, age)
    if kind == 'cat':
        print("name = ", name)

dog Ruby 7
name =  Nero
dog John 5


In [27]:
if 'dog' in animals[0][0]:
    print(animals[0][1])

Ruby


In [28]:
if animals[2][1] is 'John':
    print("John's age =", animals[2][2])

John's age = 5


## 8-3. 논리연산 (Logical Operation)

    True  and True  --> True   (1 * 1 = 1)
    True  or  True  --> True   (1 + 1 = 2)
    True  and False --> False  (1 * 0 = 0)
    True  or  False --> True   (1 + 0 = 1)
    False and False --> False  (0 * 0 = 0) 
    False or  False --> False  (0 + 0 = 0)
    
    not True  --> False
    not False --> True

In [29]:
True and True

True

In [30]:
not False

True

In [31]:
not True

False

In [32]:
x = 20
if x > 0 and x < 10:
    print(x > 0 and x < 10)
else:
    print(x > 0 and x < 10)

False


In [33]:
seasons = ['spring','summer','autumn','winter']

if seasons[0] == 'spring' and seasons[3] != 'summer':
    print(seasons[3] != 'summer')

True


In [34]:
if seasons[0] == 'spring' or seasons[1] == 'year':
    print('True')

True


In [35]:
(((2 < 3) and (3 > 0)) or ((5 > 6) and not(4 < 2)))

True

## 연습문제

1) 다음 list 에서 10 과 30 사이의 숫자 중 홀수만 골라 출력하라.

    xlist = [-5, 5, 10, 12, 13, 14, 15, 25, 30, 40, 55, 100]
    
    (결과)  
    13  15  25

2) 파이썬 강좌의 수강생 목록은 다음과 같다. 어떤 사람이 수강생 목록에 존재하는지 check 하는 함수를 작성하라.
   목록에 존재하면 True, 존재하지 않으면 False 를 반환한다.
   
   
       학생부 : ["김철수", "홍길동", "Jone Doe", "김정은", "트럼프", "성춘향"]
       
       (ex)  
       print(check_list(students, "홍길동"))  -> True

3) 다음의 주민번호 리스트에서 남, 녀 별로 90 년생 이후 출생자를 골라내라.

In [ ]:
id_list = ['920801-1041798', '800902-2048746', 
            '971010-1023987', '871203-2014987',
            '820801-1041798', '900902-2048746', 
            '941010-1023987', '971203-2014987']

man = []
lady = []

for id in id_list:
    if id[:2] >= '90' and id[7] == '1':
        
        #--- YOUR CODE HERE--------

print('남성 =', man)
print('여성 =', lady)

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 9. range( ) 함수

#### Python 은 range( ) 함수를 built-in 으로 제공한다.
#### range( ) 함수는 파라미터 (parameter) 로 받은 숫자 범위의 연속된 정수를 반환함.
#### 문법:
    range([start,] stop [,increment]) 
#### list( ) 함수는 range( ) 함수의 반환값을 list 로 변환해 줌.
#### for-loop 문에서 range( ) 함수는 list 를 반환한다.

In [2]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [3]:
list(range(0, 10, 1))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [4]:
list(range(1, 10, 2))

[1, 3, 5, 7, 9]

In [5]:
list(range(2, 10, 2))

[2, 4, 6, 8]

In [6]:
list(range(5, 5))

[]

In [7]:
list(range(5, 0, -1))

[5, 4, 3, 2, 1]

In [8]:
for i in range(5):
    print(i)

0
1
2
3
4


In [9]:
for i in range(5, -1, -1):
    print(i)

5
4
3
2
1
0


In [10]:
order = ['1st', '2nd', '3rd', '4th', '5th']
for i in range(len(order)):
    print(i, order[i])

0 1st
1 2nd
2 3rd
3 4th
4 5th


In [11]:
for i in range(len(order)-1, -1, -1):
    print(i, order[i])

4 5th
3 4th
2 3rd
1 2nd
0 1st


In [12]:
reversed(order)

In [13]:
list(enumerate(reversed(order)))

[(0, '5th'), (1, '4th'), (2, '3rd'), (3, '2nd'), (4, '1st')]

In [14]:
for i, ord in enumerate(reversed(order)):
    print(i, ord)

0 5th
1 4th
2 3rd
3 2nd
4 1st


In [15]:
for i in range(0, len(order), 2):
    print(i, order[i])

0 1st
2 3rd
4 5th


In [16]:
sum = 0
for i in range(1, 11):
    sum = sum + i
print(sum)

55


In [17]:
for i in range(5):
    print("Hello, world!")

Hello, world!
Hello, world!
Hello, world!
Hello, world!
Hello, world!


### tuple 과 for-loop


In [18]:
xtuple = (1, 'two', 2)

In [19]:
for e in xtuple:
    print(e)

1
two
2


In [20]:
for i in range(len(xtuple)):
    print("t[",i,"] = ", xtuple[i])

t[ 0 ] =  1
t[ 1 ] =  two
t[ 2 ] =  2


## enumerate() 함수

In [21]:
list(enumerate(reversed(order)))

[(0, '5th'), (1, '4th'), (2, '3rd'), (3, '2nd'), (4, '1st')]

In [22]:
for i, ord in enumerate(reversed(order)):
    print(i, ord)

0 5th
1 4th
2 3rd
3 2nd
4 1st


## 연습문제

1) &nbsp; $\sum_{k=1}^{n}k$ 을 계산하는 함수 sigma(n) 을 작성하라. (n 은 정수)

2) 다음 문장 수행 후의 output 은 ?  

```
    s = ("a", "b", "c")  
    for i in range(1, len(s) + 1):  
        sub = ""  
        for j in range(i):  
            sub = s[j] + sub  
        print(sub)
```

3) 다음 문장 수행 후의 output 은 ?
    
    def sum_part(xlist, n):
        sum = 0
        for x in xlist[n]:
            sum = sum + x
        return sum
        
    ylist = [[1, 2], [3, 4], [5, 6], [7, 8]]
    x = sum_part(ylist, 2)
    print(x)

4) 숫자로 피라미드 만들기 : 홀수 숫자를 입력으로 받아서 좌우 대칭되는 피라미드형태로 출력 한다.  


- **(알고리즘)** pyramid 를 만들어야 하므로 입력 받은 숫자 만큼 for loop n 회 반복 $\rightarrow$ for i in range(n): 이용


    - 좌우 대칭을 만들기 위해 처음 시작은 (n-i) 개의 space를 출력하고, 
    - 줄 바꿈 없이 1 에서 i+1 까지의 숫자를 출력한 후,  
    - 줄 바꿈 없이 i 에서 1 까지 역순으로 출력
    - 줄을 바꾸어 위 process를 반복
```
       1
      121
     12321
    1234321
   123454321
  12345654321
 1234567654321
 ```

In [ ]:
def pyramid(n):
    for i in range(n):
        print("  " * (n-i), end='')  #(n-i) 개의 space를 출력
        for j in range(?, i+?):        #1 에서 i+1 까지의 숫자를 출력
            print(j, end='')
        for j in range(?, ?, -1):     #i 에서 1 까지 역순으로 출력
            print(j, end='')
        print()

pyramid(7)

5) 임의의 범위의 숫자를 모두 곱하는 함수를 작성하라.
    
       ex) multiply(2,4) ==> 2 * 3 * 4 = 24

6) 숫자로 이루어진 list 의 평균을 구하는 함수를 작성하라. 단, built-in 함수를 이용하지 않고 for loop 을 이용하여 새로운 함수 작성.
   
       ex) average([2,3,4,5,6]) => 4.0

7) x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] 을 for 문을 이용하여,
```
    [1,2,3] , 4   
    [2,3,4] , 5  
    [3,4,5] , 6  
    [4,5,6] , 7  
    [5,6,7] , 8  
    [6,7,8] , 9  
    [7,8,9] , 10 을 출력하라
```

**(Hint)** [7, 8, 9], 10 을 만들기 위해 뒤에서 4 개를 남겨야 하므로 range 함수의 범위를 range(len(x) - 3) 으로 한다. 따라서,  
```python
            for i in range(len(x)-3):
                  print(x[?: ?], ',', x[?])
```

### 도전문제

    피보나치 수열(Fibonacci Sequence)을 계산하는 프로그램도 파이썬으로 간단히 작성할 수 있다. 

    피보나치 수열은 0 과 1 로 시작하고 다음의 숫자는 이전 숫자 두개를 더한 숫자들로 이루어 진다.
    0, 1, 1, 2, 3, 5, 8, 13 ......
    
$$ F_n = \begin{cases} 0 \quad \quad \quad \quad \quad \text {n = 0} \\ 1 \quad \quad \quad \quad \quad \text {n = 1} \\ F_{n-1} + F_{n-2}\quad \text {n > 1}\end{cases} $$

    n 개의 숫자로 이루어진 피보나치 수열을 출력하는 함수를 작성하라.
    
    (결과) fib(10) => [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    
**(알고리즘)**
- old, new 두개의 변수 이용 -> 초기값 : old = 0, new = 1  
- loop n times:
    - next 숫자는 old + new  
    - new 값이 old 가 되고 next 숫자가 new 값이 된다.
- loop 후 마지막 new 값이 n 번째 피보나치 수이다.

In [29]:
def fib(n):
    old, new = 0, 1
    for _ in range(n):
        # your code here 
    return new

lst = []
for i in range(7):
    lst.append(fib(i))
    
print(lst)

[1, 1, 2, 3, 5, 8, 13]


<br>
<br>
<br>
<br>
<br>

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 10. 딕셔너리 (Dictionary) 자료형

#### dictionary 는 key-value 쌍으로 이루어진 data type 이다.

#### dictionary 는 { } 내에 key: value 로 표시한다.

#### list, tuple 의 elemtent 들은 indexing 에 의해 접근하지만 , dictionary 는 key 값으로 접근한다.

    dictionary = {key: value, key: value...}
    
<img src="./images/dict1.png" width="500">

## 10.1 Dictionary basics

### dictionary 생성

In [1]:
d = {'a': 1, 'b': 2, 'c': 3}

In [2]:
type(d)

dict

### dictionary 값(value) 접근

In [5]:
d['b'] + d['c']

5

In [6]:
my_info = {'name': 'Kim', 'age': 25, 'height': 175, 
           'home': 'Seoul'}

if my_info['age'] < 30 and my_info['height'] > 170:
    print("My name is " + my_info['name'] +
            ' and I live in ' + my_info['home'] + '.')

My name is Kim and I live in Seoul.


### Dictionary 값 update

In [7]:
my_info['name'] = 'Jane'
my_info['home'] = 'New York'

In [8]:
if my_info['age'] < 30 and my_info['height'] > 170:
    print("My name is " + my_info['name'] + 
            ' and I live in ' + my_info['home'] + '.')

My name is Jane and I live in New York.


### delete dictionary key

In [37]:
my_info.pop('age')

25

In [38]:
my_info

{'name': 'Jane', 'height': 175, 'home': 'New York'}

In [41]:
del my_info['age']

KeyError: 'age'

In [40]:
del my_info['home']

In [42]:
my_info

{'name': 'Jane', 'height': 175}

### get mothod 
- dictionary.get(key, default value)

In [1]:
words = ["나", "단어", "가", "다", "가", "가", "마", 
         "다", "하", "단어", "글자"]

d = dict()

for w in words:
    d[w] = d.get(w, 0) + 1

print(d)

{'나': 1, '단어': 2, '가': 3, '다': 2, '마': 1, '하': 1, '글자': 1}


### dictionary 와 for-loop

In [10]:
age_info = {'Kim': 50, 'Lee': 20, 'Park': 15, 'Choi': 45}

In [11]:
for key in age_info:
    if key == 'Kim':
        print(key, age_info[key])

Kim 50


In [12]:
for key in age_info.keys():
    print(key)

Kim
Lee
Park
Choi


In [13]:
for value in age_info.values():
    print(value)

50
20
15
45


In [14]:
for key, value in age_info.items():
    print(key, value)

Kim 50
Lee 20
Park 15
Choi 45


In [15]:
for key in age_info.keys():
    print(age_info[key])

50
20
15
45


In [16]:
for key in age_info.keys():
    print(key, age_info[key])

Kim 50
Lee 20
Park 15
Choi 45


In [17]:
for key, value in age_info.items():
    print(key, '\t\t', value)

Kim 		 50
Lee 		 20
Park 		 15
Choi 		 45


### dictionary 정렬

- key 순으로 정렬

In [18]:
d = {'c': 1, 'a': 2, 'b': 3}
sorted(d)

['a', 'b', 'c']

In [19]:
for key, value in d.items():
    print(key, '\t', value)

c 	 1
a 	 2
b 	 3


In [20]:
for key in sorted(d):
    print(key, '\t', d[key])

a 	 2
b 	 3
c 	 1


- value 순으로 정렬  
- without lambda 

In [21]:
tmp = list()

for k, v in d.items():
    tmp.append((v, k))
    
tmp.sort()
tmp

[(1, 'c'), (2, 'a'), (3, 'b')]

- with lambda

In [22]:
sorted(d.items(), key=lambda kv: kv[1])

[('c', 1), ('a', 2), ('b', 3)]

### dictionary 에 key 존재 유무 check

In [23]:
d = {'a': 1, 'b': 2, 'c': 3}
'a' in d

True

In [24]:
'd' in d

False

### dictionary 를 이용한 list 의 중복 element count

In [25]:
def word_count_dict(word_list):
    word_count = {}
    for word in word_list:
        word = word.lower()
        if not word in word_count:
            word_count[word] = 1
        else:
            word_count[word] += 1
    return word_count

In [26]:
word_list = ['test', 'Test', 'apple', 'Apple', 
             'orange', 'dog', 'caT', 'CAT']

words_cnt = word_count_dict(word_list)

In [27]:
words_cnt

{'test': 2, 'apple': 2, 'orange': 1, 'dog': 1, 'cat': 2}

In [28]:
for k, v in words_cnt.items():
    print(k, v)

test 2
apple 2
orange 1
dog 1
cat 2


### Simplified word_count_dict

- get(key, default value) method 를 이용하여 위의 예제를 단순화

In [2]:
def word_count_simple(word_list):
    word_count = {}
    for word in word_list:
        word = word.lower()
        word_count[word] = word_count.get(word, 0) + 1
    return word_count

word_list = ['test', 'Test', 'apple', 'Apple', 'orange', 
             'dog', 'caT', 'CAT']

word_count_simple(word_list)

{'test': 2, 'apple': 2, 'orange': 1, 'dog': 1, 'cat': 2}

### 계층적 자료구조 구성

In [30]:
stock = {
    'name': '삼성전자',
    'market': 'KOSPI',
    'close': [
        ('2010-08-01', 45000),
        ('2010-08-02', 47000),
        ('2010-08-03', 42000),
    ],
    'market-cap': 2919097,
    'PER': 15.17
}

stocks = {
    'sec': {
        'name': '삼성전자',
        'market': 'KOSPI',
        'close': [
            ('2010-08-01', 45000),
            ('2010-08-02', 47000),
            ('2010-08-03', 42000),
        ],
        'market-cap': 2919097,
        'PER': 15.17
    },
    'lg' : {
        'name': 'LG',
        'market': 'KOSPI',
        'close': [
            ('2010-08-01', 45000),
            ('2010-08-02', 47000),
            ('2010-08-03', 42000),
        ],
        'market-cap': 2919097,
        'PER': 15.17
    },
    'TESLA' : {
        'name': 'TESLA',
        'market': 'KOSPI',
        'close': [
            ('2010-08-01', 45000),
            ('2010-08-02', 47000),
            ('2010-08-03', 42000),
        ],
        'market-cap': 2919097,
        'PER': 15.17
    }
}

In [31]:
stocks['lg']

{'name': 'LG',
 'market': 'KOSPI',
 'close': [('2010-08-01', 45000),
  ('2010-08-02', 47000),
  ('2010-08-03', 42000)],
 'market-cap': 2919097,
 'PER': 15.17}

In [32]:
for c in stock['close']:
    print(c[0], c[1])

2010-08-01 45000
2010-08-02 47000
2010-08-03 42000


In [33]:
stock['phone'] = '02-3455-2020'

In [34]:
stock

{'name': '삼성전자',
 'market': 'KOSPI',
 'close': [('2010-08-01', 45000),
  ('2010-08-02', 47000),
  ('2010-08-03', 42000)],
 'market-cap': 2919097,
 'PER': 15.17,
 'phone': '02-3455-2020'}

In [35]:
stock.keys()

dict_keys(['name', 'market', 'close', 'market-cap', 'PER', 'phone'])

### 연습문제

1) 다음의 출력 결과는 ?

    d = {'a': 2, 'b': 4, 'c': 9}
    for x in sorted(d):
        print(d[x], end="")

In [1]:
d = {'a': 2, 'b': 4, 'c': 9}
for x in sorted(d):
    print(d[x], end="")

249

2) 다음의 출력 결과는 ?

    d = {'a': 2, 'b': 4, 'c': 9}
    for x in sorted(d.values()):
        print(x, end="")

In [2]:
d = {'a': 2, 'b': 4, 'c': 9}
for x in sorted(d.values()):
    print(x, end="")

249

3) 다음의 출력 결과는 ?

    d = {'a': 21, 'b': 4, 'c': 9}
    for x in sorted(d.items()):
        print(x, end="")

In [3]:
d = {'a': 21, 'b': 4, 'c': 9}
for x in sorted(d.items()):
    print(x, end="")

('a', 21)('b', 4)('c', 9)

4) dictionary 를 이용하여 자신의 정보를 구조화 하여 작성  

        ex) {'name': '오영제', 
              '성별': '남성', 
              '거주지': {'도시': '서울', '동': '충무로', '주소': '123-45'}, 
              '전화번호': 1012345678, 
              '신장': 175.17}

5) 두개의 주사위를 던져서 두 주사위의 합이 같은 것끼리 출력하라. 두 주사위의 합을 key value 로 사용하여 dictionay 를 생성한다.

- pseudo-code

```
    d = {}
    
    Loop with 주사위1 from 1 to 6
        Loop with 주사위2 from 1 to 6
            newTuple = (주사위1, 주사위2)
            added = 주사위1 + 주사위2
            if added 가 d 에 없으면 empty list 를 d 에 추가
            d 의 기존 list 에 append
```

In [7]:
import pprint
d = {}

for dice1 in range(1, 7):
    for dice2 in range(1, 7):
        
        newTuple = (dice1, dice2)
        added = dice1 + dice2
        
        if added not in d:
            d[added] = []
        d[added].append(newTuple)

pprint.pprint(d)

{2: [(1, 1)],
 3: [(1, 2), (2, 1)],
 4: [(1, 3), (2, 2), (3, 1)],
 5: [(1, 4), (2, 3), (3, 2), (4, 1)],
 6: [(1, 5), (2, 4), (3, 3), (4, 2), (5, 1)],
 7: [(1, 6), (2, 5), (3, 4), (4, 3), (5, 2), (6, 1)],
 8: [(2, 6), (3, 5), (4, 4), (5, 3), (6, 2)],
 9: [(3, 6), (4, 5), (5, 4), (6, 3)],
 10: [(4, 6), (5, 5), (6, 4)],
 11: [(5, 6), (6, 5)],
 12: [(6, 6)]}


<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 11. 스트링 (String) 

## 11-1. 스트링 (String) 자료형과 string 함수

### Python 의 string 자료형은 text 형태의 문자열을 말함.
### 문자열은 " " , ' '  혹은 """ """ 로 둘러싸인 문자들의 집합임.
    "Life is short"
    'a'
    "123"
    """String is Python's text type."""
### string 자료는 수정할 수 없음 (immutable)

* 수정할 수 있는 자료형 (mutable)  
    list
    
* 수정할 수 없는 자료형 (immutable)  
    string, tuple
    
### string 간의 +, * 연산이 가능

### str( ) 내장함수로 string type을 변환

### string 은 iterable 한 특성을 가지므로 list 와 같이 for-loop, len( ), indexing, slicing이 가능

### string 과 list 비교

In [3]:
xlist = ['g','o','o','d']
xlist[1] = 'x'

In [4]:
xlist

['g', 'x', 'o', 'd']

In [5]:
s1 = 'good'
s1[1] = 'x'

TypeError: 'str' object does not support item assignment

### string operation

In [6]:
s2 = "Good" + " morning"
print(s2)

Good morning


In [7]:
s3 = "good" * 3
print(s3)

goodgoodgood


In [8]:
s4 = "Number" + 3

TypeError: can only concatenate str (not "int") to str

In [9]:
s4 = "Number" + str(3)
s4

'Number3'

In [10]:
s5 = "Python is powerful"
print(s5[0], s5[-1], s5[:3])

P l Pyt


### slicing

In [11]:
s5[-1:-4:-1]

'luf'

In [12]:
s5[::-1]

'lufrewop si nohtyP'

In [13]:
len(s5)

18

### string and for-loop

In [14]:
for ch in s5:
    print(ch, end="")

Python is powerful

In [15]:
s1 + s2 + s3 + s4 + s5

'goodGood morninggoodgoodgoodNumber3Python is powerful'

In [16]:
s1 = """String is Python's "text" type."""

In [17]:
print(s1)

String is Python's "text" type.


In [18]:
my_string = "aeiou"
if "a" in my_string and "u" in my_string:
    print("Both conditions are True")

Both conditions are True


## 11-2. Escape Sequence 

### "\\" 로 시작하며 "\\"에 뒤따르는 문자에 특별한 의미를 부여하기 위해 사용
    예) \n : new line
       \\  : backslash ('\')
       \', \" : 따옴표 (quotation mark)
       \t  : tap

In [19]:
print("First Line \nSecond Line")

First Line 
Second Line


In [20]:
print("Seoul-Busan \\non-stop")

Seoul-Busan \non-stop


In [21]:
print('He\'s my firend')

He's my firend


In [22]:
print("He's my friend")

He's my friend


In [23]:
print("She said \"I love him.\"")

She said "I love him."


In [24]:
print("tap character \tmakes spaces")

tap character 	makes spaces


In [25]:
print("tap character \t\tmakes spaces")

tap character 		makes spaces


In [26]:
s1 = """I didn't know what was "it"."""
print(s1)

I didn't know what was "it".


In [27]:
s = "I was bo\nrn in Seoul and I liv\ted in Seoul."
print(s)

I was bo
rn in Seoul and I liv	ed in Seoul.


- escape sequence 의 길이는 1

In [28]:
x = 'a\nb'
y = 'a\tb'
print(x)
print(y)
len(x), len(y)

a
b
a	b


(3, 3)

### 연습문제

1) parameter 로 받은 string 의 양 끝단 2 글자를 붙여서 반환하는 함수 작성
- string 의 slicing 및 concatenation 이용
```Python
    def both_ends(s):
        if len(s) < 2:
            return ''
        first2 = s[:?]
        last2 = s[?:]
        return first2 + last2

    print(both_ends('spring'))  #spng 
    print(both_ends('Hello'))    #Helo
```  

2) 두개의 문자열을 parameter 로 받아 서로의 첫번째 두 글자를 교환한 후 중간에 한칸 띄우고 반환하는 함수 작성

```python
    def mix_up(a, b):
        a_swapped = b[:?] + a[?:]
        b_swapped = a[:?] + b[?:]
        return ? + ' ' + ?

    print(mix_up('frog', 'dinner'))   # diog frnner
```

3) 두개의 parameter 를 받아서 두개의 시작이 같으면 SAME 다르면 Different 를 반환하는 함수 작성
```python
    test('a', 'a')     # SAME
    test('aba', 'cba')   # Different
```

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 12. String 의 여러가지 사용방법

## 12-1. string class 의 주요 메소드 (method)

### 대, 소문자 변환
    lower( ), upper( )
### 문자열(string) 내의 특정 문자 갯수 
    count( )
### 앞부분 / 뒷부분 화이트 스페이스(whitespace) 제거
    
    Python 의 whitespace
    
    ' ' : space (공백)
    '\n': newline (줄바꿈)
    '\r': carriage return (
    '\t': 수평탭 (horizontal tab)
    '\v': 수직탭 (vertical tab)
    
    strip( )  - 앞, 뒤의 whitespace 제거
    lstrip( ) - 앞부분의 whitespace 제거
    rstrip( ) - 뒷부분의 whitespace 제거
    
    
### 문자열(string) 내의 특정 문자, 문자열의 위치
    find( ), index( )
    
### 문자열을 다른 문자열로 대치 
    replace( )
    
### 문자열 쪼개기 / 합치기
    split( )
    join( )

### built-in methods

In [1]:
s = "good morning"

In [2]:
dir(s)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',


In [3]:
S = s.upper()
print(S)

GOOD MORNING


In [4]:
S.lower()

'good morning'

In [5]:
s.count('o')

3

In [6]:
w = "   good morning    "

In [7]:
print(w.strip())

good morning


In [8]:
w = "   good \n morning   "

In [9]:
w.strip()

'good \n morning'

In [10]:
w.lstrip()

'good \n morning   '

In [11]:
w.rstrip()

'   good \n morning'

In [12]:
s = "World Cup"

In [13]:
s.find("C")

6

In [14]:
s.index("Cup")

6

In [15]:
s.find("Cup")

6

In [16]:
s.find("k")

-1

In [17]:
s.index("k")

ValueError: substring not found

In [18]:
try:
    s.index("k")
except:
    print("not found")

not found


In [19]:
s.replace("World", "bad")

'bad Cup'

In [20]:
slist = s.split()

In [21]:
print(slist)

['World', 'Cup']


In [22]:
s = 'World and Cup'
s.split("and")

['World ', ' Cup']

- 구분자를 이용한 list 의 단일 string 변환

In [23]:
' '.join(slist)

'World Cup'

In [24]:
'-'.join(["one", "to", "one"])

'one-to-one'

## 12-2. 문자열 포매팅 (String Formatting)

- 문자열 포매팅이란 문자열 내에 어떤 값을 삽입하는 것


-  { } 로 위치를 표시하고 format( ) method 로 mapping 시킨다.


-  `{[순서]:[표시형식]}` 지정 가능


-  format( ) method 대신 % 를 이용할 수도 있다. $\rightarrow$ Python 2.7 방식

In [25]:
"Python is {}".format("powerful")

'Python is powerful'

In [26]:
"Python is {} and {}".format("easy", "powerful")

'Python is easy and powerful'

In [27]:
"Float : {}, List: {}".format(1.5, ['a','b','c'])

"Float : 1.5, List: ['a', 'b', 'c']"

In [28]:
print("많이 팔리는 순서는 {0}, {1}, {2} 이고 {1} 는 품절입니다."
          . format("사과","바나나","망고"))

많이 팔리는 순서는 사과, 바나나, 망고 이고 바나나 는 품절입니다.


In [29]:
print("Friends are {s}, {b} and {j}.".
      format(j="Bill", b="John", s="Sean"))

Friends are Sean, John and Bill.


In [30]:
"{}-{}-{}".format(5, 10, 15)

'5-10-15'

In [31]:
"{0:5}-{1:5}-{2:5}".format(5, 10, 15)

'    5-   10-   15'

In [32]:
"{0:<10} left justify".format("LEFT")

'LEFT       left justify'

In [33]:
"{:>10} right justify".format("RIGHT")

'     RIGHT right justify'

In [34]:
"{:^10} center justify".format("CENTER")

'  CENTER   center justify'

In [35]:
"{:*^10} filling with *".format("CENTER")

'**CENTER** filling with *'

In [36]:
y = 3.141592
"{0} is {1:0.4f}".format("Pi", y)

'Pi is 3.1416'

In [37]:
import math
r2 = math.sqrt(2)
print("square root 2 is {:.4f}".format(r2))

square root 2 is 1.4142


In [38]:
print("square root 2 is {:10.4f}".format(r2))

square root 2 is     1.4142


In [39]:
print("square root 2 is [{:10.4f}]".format(r2))

square root 2 is [    1.4142]


In [40]:
print("square root 2 is {{{:10.4f}}}".format(r2))

square root 2 is {    1.4142}


## 12-3. 포맷문자열 리터럴 (f-문자열)

-  f+string 내에 {expression}  사용

In [41]:
import math
print(f'The value of pi is approximately {math.pi:.3f}.')

The value of pi is approximately 3.142.


In [42]:
table = {'Sjoerd': 4127, 'Jack': 4098, 'Dcab': 7678}

for name, phone in table.items():
    print(name, '==> {}'.format(phone))
    print(f'{name:10} ==> {phone:10d}')

Sjoerd ==> 4127
Sjoerd     ==>       4127
Jack ==> 4098
Jack       ==>       4098
Dcab ==> 7678
Dcab       ==>       7678


### Python 2.7 형식

In [43]:
"I have %s apples." % "three"

'I have three apples.'

In [44]:
"I have %d apples" % 5

'I have 5 apples'

In [45]:
"square root is %.4f" % r2

'square root is 1.4142'

In [46]:
"%% formatting is %s python %s" % ("old", "style")

'% formatting is old python style'

### 연습문제

1) 다음의 print 결과는 ?  

    s = "Python is awesome"  
    print(s[1:3])

2) "Spam and Eggs" 를 입력할 때 다음 프로그램의 출력 결과는 ?  

    def main():
        msg = input("Enter a phrase:  ")  
        for w in msg.split():  
            print(w[0], end="")  

    main()
    
    (결과) SaE

3) 다음의 출력 결과는 ?

    for x in "Mississippi".split("i"):
        print(x, end="")
        
        (결과) Msssspp

4) 다음 출력 결과는 ?

    s = "Jane Doe"  
    print(s[3 : 1 : -1])
    
    (결과) en

5) s = "Hello, Python World" 을 알파벳 별로 몇개인지 계산 (단, 대소문자 무시)

* hint : 
    1. "," 와 " " 을 "" 로 replace  
    2. s.lower() 를 이용하여 소문자로 통일
    3. dictionary 를 이용하여 알파벳 별로 개수 누적

(결과)  
```
        d 1
        e 1
        h 2
        l 3
        n 1
        o 3
        p 1
        r 1
        t 1
        w 1
        y 1
```

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 13. Module

#### 미리 만들어 놓은 Python 프로그램들을 모듈(Module) 혹은 package 라고 한다.

#### module 은 python 함수를 모아 놓은 것이다.

#### module 은 import 명령어로 불러온다.

#### 관련된 기능의 module  여러개를 한 덩어리로 묶은 것을 package 라고 부른다. package 는 기본 Python 3.7 에 포함되어 있지 않고 필요할 때 추가로 설치해야 한다.

#### 별도 설치가 필요 없는 Python 모듈을 built-in module (내장모듈) 이라고 부른다.

#### 내장 module 의 예:
    math, random
    
#### import 문 문법:
    import <모듈명>
    from <모듈명> import <함수명, *>
    from <모듈명> import <함수명, *> as <별명>

## 13-1. 전체 module 불러오기

In [1]:
import math

In [2]:
type(math)

module

### module 내에 포함된 함수들

In [3]:
dir(math)

['__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'pi',
 'pow',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc']

In [4]:
help(math.sin)

Help on built-in function sin in module math:

sin(x, /)
    Return the sine of x (measured in radians).



In [5]:
math.exp(2)

7.38905609893065

In [6]:
help(math.exp)

Help on built-in function exp in module math:

exp(x, /)
    Return e raised to the power of x.



In [7]:
help(math.log)

Help on built-in function log in module math:

log(...)
    log(x, [base=math.e])
    Return the logarithm of x to the given base.
    
    If the base not specified, returns the natural logarithm (base e) of x.



In [8]:
math.log(2)

0.6931471805599453

In [9]:
math.floor(3.14)

3

In [10]:
math.ceil(3.14)

4

### 전체 module * 로 불러오기

In [11]:
from math import *

In [12]:
help(sin)

Help on built-in function sin in module math:

sin(x, /)
    Return the sine of x (measured in radians).



In [13]:
help(radians)

Help on built-in function radians in module math:

radians(x, /)
    Convert angle x from degrees to radians.



In [14]:
radians(0)

0.0

In [15]:
radians(90)

1.5707963267948966

In [16]:
math.sin(radians(0))

0.0

In [17]:
math.sin(radians(90))

1.0

### 개별 함수 불러오기

In [18]:
from math import sqrt, pi

In [19]:
sqrt(4)

2.0

In [20]:
print(pi)

3.141592653589793


In [21]:
from math import sin as sine

In [22]:
sine(radians(0))

0.0

In [23]:
sine(radians(90))

1.0

In [24]:
from collections import Counter

lst = ['cat', 'dog', 'monkey', 'cat', 'dog', 'cat']

counts = Counter(lst)
counts

Counter({'cat': 3, 'dog': 2, 'monkey': 1})

In [31]:
counts['cat']

3

In [32]:
counts.get('cat')

3

### 사용자가 작성한 module 불러오기

- module_test.py file 내의 circle 함수 call

```
    def circle(pi, radius):
        print("pi = ", pi)
        print("반지름 = ", radius)
        return pi * radius ** 2

    def calculate_volume(length, width, depth):
        return length * width * depth
```

In [26]:
from module_test import circle

In [27]:
circle(3.14, 10)

pi =  3.14
반지름 =  10


314.0

### 상위 folder 의 module 불러오기

- 상위 folder 의 boilerplate.py 파일

```
import sys

def hello():
    print('Hello Python')

if __name__ == '__main__':
    hello()
    print(sys.argv)
    print(sys.argv[0])
    print(sys.argv[1])
```

In [36]:
import sys
sys.path.append('../')

In [37]:
import boilerplate

In [38]:
boilerplate.hello()

Hello Python


##  연습문제

### boilerplate.py 프로그램을 terminal 에서 실행

- sys.argv 파라미터를 출력하여 내용 이해

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 14. 클라스(Class)

## 14-1. 클라스와 메소드 (Class 와 Method)

### Python 의 class 와 method 는 C++, Java 등의 객체지향 프로그래밍의 개념과 동일

### 객체지향개발(OOP - Object Oriented Programming) 4가지 특성
- 객체지향개발(Object Oriented Programming)의 특성은 크게 추상화, 캡슐화, 상속성, 다형성이 있다.

#### 1. 추상화(Abstraciton)
- 공통의 속성이나 기능을 묶어 이름을 붙이는 것
- OOP 에서 클래스를 정의하는 것을 추상화라고 할 수 있다.

#### 2. 은닉화 (Encapsulation)
- 변수와 함수를 하나로 묶어서 외부에서의 접근을 막고, 함수를 통해서만 접근 가능하도록 하는 것

#### 3. 상속 (Inheritance)
- 상위 개념의 특징과 메소드를 하위 개념이 물려 받아 재사용  

#### 4. 다형성 (Polymorphism)
- 부모 클래스로부터 물려받은 가상 함수를 자식 클래스 내에서 오버라이딩 하여 사용하는 것

### Python 에서는 모든 것을 object 라고 부름
    data (string, integer, float), function, built-in function
    
### Python 에서 모든 object 는 특정 class 에 속함

### class 란 객체 (object) 를 정의하는 틀 혹은 설계도와 같은 의미

### method 는 class 내에서 정의한 함수임. class 는 속성 (data) 와 메소드(method) 로 구성

    class ClassName:
        <data>
        
        def methodName(self):
            <method body>
            return
            
### dir( ) 내장함수는 객체가 어떤 변수와 메소드를 가지고 있는지 반환
    Python 언어 자체에서 사용하는 특별한 method 들은 __name__ 형식의 이름을 가지고 있음

In [8]:
print(type("string"))

<class 'str'>


In [9]:
print(type(15))

<class 'int'>


In [10]:
print(type(12.5))

<class 'float'>


In [11]:
def f(n):
    return n

In [12]:
print(type(f))

<class 'function'>


In [13]:
print(type(print))

<class 'builtin_function_or_method'>


In [15]:
class Car:
    color = "blue"
    max_speed = 100
    
    def __init__(self):
        self.max_speed = 50
        print("Car Instance 생성")
    
    def speedCheck(self, n):
        if n > self.max_speed:        # self 생략하면 error
            print(self.max_speed)
            print("Too fast")
        else:
            print("Good speed")

In [16]:
sonata = Car()

Car Instance 생성


In [17]:
sonata.speedCheck(100)

50
Too fast


In [14]:
sonata.max_speed

500

In [18]:
print(type(sonata))

<class '__main__.Car'>


In [19]:
print(type(sonata.speedCheck))

<class 'method'>


## 14-2 Python 의 Object Oriented Programming

- Class : 속성(attribute) 과 method 로 구성
    
    class Classname:
        def __init__(self, *args):
            self.변수명 (속성 정의)
            
        def methods(self, *args):
            메소드 정의
       
- Constructor (생성자)  : __init__()  


    class 의 instance 생성 시 자동으로 호출되어 object 를 초기화 하는 method
    
    
            
- Inheritance (상속)

    class Subclass(superclass):
    
        
- Encapsulation (은닉화)  

    self._변수명
    
    
- Polymorphism (다형성)

    Python 은 자체로 다형성의 특성을 가진 언어이다.
    
   1 + 2 ==> 3  
   '1' + '2'  =='12'

- Python interpreter 는 + operator 를 만나면 `__add__` method 호출

In [20]:
a = 1
a.__add__(2)    # numeric 연산

3

In [21]:
a = '1'
a.__add__('2')    # string 연산

'12'

### Class example

In [22]:
# class 정의
class Person:
    def __init__(self, name):
        self.name = name
        
    def sayHello(self):
        print(self.name, "님 안녕하세요")
        
# instance 생성
A = Person("파이썬")

# instance 를 통한 method 호출
A.sayHello()

# instance 를 통한 attribute 접근
print(A.name)

파이썬 님 안녕하세요
파이썬


### class 변수와 instance 변수

In [23]:
class Person:
    country = 'Korea'
    def __init__(self, country):
        self.country = country
        
A = Person('USA')

print(Person.country)     # class 변수
print(A.country)             # instance 변수

Korea
USA


### Inheritance (상속)

In [33]:
class Korean:        # parent class
    def __init__(self, name):
        self.name = name
        
    def speak(self):
        print('나 {}는 한국말을 합니다.'.format(self.name))
    
    def sex(self):
        print('부모 성별')
        
class Man(Korean):      # child class
    def sex(self):
        print('나 {}는 한국 남자 입니다.'.format(self.name))
        
class Woman(Korean):  # child class
    def sex(self):
        print('나 {}는 한국 여자 입니다.'.format(self.name))
        
A = Man('김철수')
A.speak()
A.sex()
print()
B = Woman('김순이')
B.speak()
B.sex()

나 김철수는 한국말을 합니다.
나 김철수는 한국 남자 입니다.

나 김순이는 한국말을 합니다.
나 김순이는 한국 여자 입니다.


## 14.3 The `__call__` method

- class 의 초기화된 인스턴스를 호출할 때마다 수행. 여러 인수를 가질 수 있으며 원하는대로 수행하도록 정의 가능.

    - 매개 변수 변경,
    - 메시지 인쇄,
    - 새 변수 생성 등

In [47]:
class MyClass:
    def __init__(self, x):
        self.x = x
    
    def __call__(self, y):
        self.x += y
        print(self.x)

In [48]:
instance_a = MyClass(10)

instance_a(20)

30


## 14.4 `__str__`, `__repr__`

- __str__ - print() 문 호출시 string 반환
- __repr__ - object 자체 출력시 readable 한 형태 반환
- __repr__ 만 정의되어 있고 __str__ 이 생략된 경우 __str__=__repr__ 와 같음

In [49]:
class Korean:        # parent class
    
    def speak(self):
        print('나는 한국말을 합니다.')

    def __str__(self):
        return "Print 문 사용시 호출"
    
    def __repr__(self):
        return "object 반환시 호출 "
    
A = Korean()

In [50]:
A.speak()

나는 한국말을 합니다.


In [51]:
A

object 반환시 호출 

In [52]:
print(A)

Print 문 사용시 호출


## 14.5 다형성 (polymorphism) 구현

**super()**

- 자식클래스에서 부모클래스의 내용을 사용하고 싶은 경우
- super().부모클래스내용

In [53]:
class Person:   
    
    class_var = '클라스변수'
    
    def __init__(self, name, age):
        self.name = name
        self.age = age
        
    def show_salary(self):
        print("Salary is unknown")

class Employee(Person):
    def __init__(self, name, age, salary):   # Person 의 __init__() override
        super().__init__(name, age)          # Person 의 __init__() 내용을 상속
        self.salary = salary
        
    def show_salary(self):          # Person 의 show_salary() override
        print( "{}'s salary is {} and age is {}"
                  .format(self.name, self.salary, self.age))
        
A = Person('Tom', 30)
A.show_salary()

B = Employee('Tom', 30, 10000)
B.show_salary()

Salary is unknown
Tom's salary is 10000 and age is 30


## 14.6 다중 상속

In [3]:
class Person:
    def greeting(self):
        print('안녕하세요.')
        
class University:
    def manage_grade(self):
        print('학점 관리')
        
class Student(Person, University):
    def study(self):
        print('공부하기')

tom = Student()
tom.greeting()
tom.manage_grade()
tom.study()

안녕하세요.
학점 관리
공부하기


## 14.7 Class 내의 method 종류

### instance method 
    - 인스턴스를 첫번째 인자로 받음
    - 관습적으로 self 로 표시


### class method 
    - 클래스를 첫번째 인자로 받음  
    - 관습적으로 cls  로 표시
    - 모든 인스턴스가 공유하는 클래스 변수와 같은 데이터를 생성, 변경 또는 참조하기 위한 메소드
    
    
### static method
    - 첫번째 인자(self) 자동 생성 되지 않음 - class 내의 일반 함수  
    - class 내의 utility 성격의 method 로 사용  
    - class, instance 에서 모두 접근 가능 

In [54]:
class Employee(object):
    
    raise_amount = 1.0  # 연봉 인상율 클래스 변수
    
    def __init__(self, first, last, pay):
        self.first = first
        self.last = last
        self.pay = pay
    
    def apply_raise(self):
        self.pay = int(self.pay * self.raise_amount)
        
    def full_name(self):
        return '{} {}'.format(self.first, self.last)
        
    def get_pay(self):
        return '현재 "{}"의 연봉은 "{}"입니다.'
    .format(self.full_name(), self.pay)
    
    # class 메소드 데코레이터를 사용하여 class 메소드 정의
    @classmethod
    def change_raise_amount(cls, amount):
        cls.raise_amount = amount
        print('인상율 "{}"가 적용 되었습니다.'.format(amount))
    
    # static 메소드 데코레이터를 사용하여 static 메소드 정의
    @staticmethod
    def calc_tax(amount, tax_rate):
        print(amount * tax_rate)

emp_1 = Employee('Sanghee', 'Lee', 50000)
emp_2 = Employee('Minjung', 'Kim', 60000)

# 연봉 인상 전
print(emp_1.get_pay())
print(emp_2.get_pay())

# 연봉 인상율 변경
# Employee.raise_amount = 1.2 와 동일한 결과
Employee.change_raise_amount(1.2)    

# 연봉 인상
emp_1.apply_raise()
emp_2.apply_raise()

# 연봉 인상 후
print(emp_1.get_pay())
print(emp_2.get_pay())

Employee.calc_tax(1000, 0.05)   # static method 도 instance 를 통해 접근
emp_1.calc_tax(1000, 0.04)

현재 "Sanghee Lee"의 연봉은 "50000"입니다.
현재 "Minjung Kim"의 연봉은 "60000"입니다.
인상율 "1.2"가 적용 되었습니다.
현재 "Sanghee Lee"의 연봉은 "60000"입니다.
현재 "Minjung Kim"의 연봉은 "72000"입니다.
50.0
40.0


## 연습문제

- Car class 를 parent class 로 만들고 Sonata 와 Volvo 를 child class 로 구성

- ``#CODE HERE`` 부분을 채워서 출력된 결과와 동일하도록 code 완성

```
black
앞으로 전진 : 시속 100 km
뒤로 후진 70 km/hour
white
5000000
앞으로 전진 : 시속 200 km
후진 불가
```

In [15]:
class Car:
    def __init__(self, color='red', power=2000):
        self.color = color
        self.power = power
        self.speed = 0

    def forward(self, speed):
        self.speed += speed
        return '앞으로 전진 : 시속 {} km'.format(self.speed)

    def backward(self):
        pass

class Sonata(Car):
    def __init__(self, color, power, size):
        
        # CODE HERE
        
    def backward(self, speed):
        
        # CODE HERE
        
class Volvo(Car):
    def __init__(self, color, power, size, price):
        
        # CODE HERE

    def forward(self, speed):
        
        # CODE HERE
        
    def backward(self):
        
        # CODE HERE
        
sonata = Sonata('black', 1800, 5)
volvo = Volvo('white', 2500, 7, 5000000)

print(sonata.color)
print(sonata.forward(100))
print(sonata.backward(30))

print(volvo.color)
print(volvo.price)
print(volvo.forward(100))
print(volvo.backward())

black
앞으로 전진 : 시속 100 km
뒤로 후진 70 km/hour
white
5000000
앞으로 전진 : 시속 200 km
후진 불가
None


<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 15. 집합(set)

### 집합에 관련된 것들을 쉽게 처리하기 위한 자료형이다.

### 집합 자료형은 다음과 같이 set 키워드 혹은 { } 를 이용해 만들 수 있다.

### 중복을 허용하지 않는다.
### 순서가 없다 (unordered) --> indexing 을 허용 않는다.
### 교집합 (&), 합집합( | ), 차집합( - )

In [1]:
s1 = set([1,2,3])
print(s1)

{1, 2, 3}


In [2]:
type(s1)

set

In [3]:
s2 = {'a', 'b', 'c'}

In [4]:
s2[1]

TypeError: 'set' object is not subscriptable

In [5]:
s1 = set([1, 2, 3, 4, 5, 6])
s2 = set([4, 5, 6, 7, 8, 9])

### 교집합

In [6]:
s1 & s2

{4, 5, 6}

In [7]:
s1.intersection(s2)

{4, 5, 6}

### 합집합

In [8]:
s1 | s2

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [9]:
s1.union(s2)

{1, 2, 3, 4, 5, 6, 7, 8, 9}

### 차집합

In [10]:
s1 - s2

{1, 2, 3}

In [11]:
s1.difference(s2)

{1, 2, 3}

### element 제거, 추가

In [12]:
s1.remove(2)

In [13]:
s1

{1, 3, 4, 5, 6}

In [14]:
s1.add('added')

In [15]:
s1

{1, 3, 4, 5, 6, 'added'}

### 여집합

In [16]:
A = {1,2}
B = {3, 4, 5}
Union = A | B
 
AC = Union.difference(A)
 
print(AC)

{3, 4, 5}


# 연습문제
## 자카드유사도 (Jaccard Similarity)

$$J(A, B) = \frac{|A\cap B|}{|A\cup B|} = \frac{|A\cap B|}{|A|+|B|-|A\cap B|}$$

### 두개의 문서간의 유사도
$$J(doc1, doc2) = \frac{doc1\cap doc2}{doc1\cup doc2} $$

In [17]:
doc1 = "apple banana everyone like likely watch card holder"
doc2 = "apple banana coupon passport love you"

tokenized_doc1 = doc1.split()
tokenized_doc2 = doc2.split()

print(tokenized_doc1)
print(tokenized_doc2)

['apple', 'banana', 'everyone', 'like', 'likely', 'watch', 'card', 'holder']
['apple', 'banana', 'coupon', 'passport', 'love', 'you']


**두 문서의 전체 단어 집합**

In [3]:
union = # CODE HERE
print(union)
print(len(union))

{'holder', 'like', 'coupon', 'everyone', 'likely', 'passport', 'apple', 'banana', 'you', 'watch', 'card', 'love'}
12


**두 문서에 공통으로 나오는 단어**

In [4]:
inter = # CODE HERE
print(inter)

{'banana', 'apple'}


In [8]:
print("두 문서간의 Jaccard 유사도 : {:.2f} ".format(len(inter)/len(union)))

두 문서간의 Jaccard 유사도 : 0.17 


<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>

# 16. while 문을 이용한 무한 반복 (indefinite loop)

#### while 문은 주어진 조건이 만족되는 동안 (True 인 동안) 무한히 반복 처리를 함.
    while <test 문>:
        <body>

#### break 문을 이용하여 반복 처리를 끝낼 수 있음. continue 문은 끝내지 않고 계속 진행.
    while True:
        if <조건 check>:
            continue
        else:
            break

#### 대부분의 while-loop 은 for-loop 으로 해결 가능하므로 while-loop 은 꼭 필요한 경우에만 사용

In [1]:
x = 0
while x < 10:
    x += 1
    print(x, end=" ")

1 2 3 4 5 6 7 8 9 10 

In [2]:
x = 0
for i in range(10):
    x += 1
    print(x, end=" ")

1 2 3 4 5 6 7 8 9 10 

In [3]:
x = 0
while True:
    x += 1
    print(x, end=" ")
    if x >= 10:
        break

1 2 3 4 5 6 7 8 9 10 

In [4]:
x = 0
while True:
    x += 1
    print(x, end=" ")
    if x < 10:
        continue
    else:
        break

1 2 3 4 5 6 7 8 9 10 

In [5]:
1368 % 180

108

## while 문 연습문제 (유클리드 호제법)

- 2개의 자연수 또는 정식의 최대공약수를 구하는 알고리즘
- 두 수가 서로 상대방 수를 나누어서 결국 원하는 수를 얻는 알고리즘
- 명시적으로 기술된 가장 오래된 알고리즘 (기원전 300 년)
- 원리
    - 2개의 자연수 a, b에 대해서 a 를 b 로 나눈 나머지를 r 이라 하면(단, a>b)
    - a 와 b의 최대공약수는 b 와 r 의 최대공약수와 같다. 
    - 이 성질에 따라, b 를 r 로 나눈 나머지 r’ 를 구하고, 다시 r 을 r’ 로 나눈 나머지를 구하는 과정을 반복
    - 나머지가 0 이 되었을 때 나누는 수가 a와 b의 최대공약수

Example 1) 1071과 1029의 최대공약수

- 1071 > 1029 이므로 1071 / 1029 $\rightarrow$ 나머지 42  
- 1029 / 42 $\rightarrow$ 나머지 21  
- 42 / 21 $\rightarrow$ 나머지 0 $\rightarrow$ 21 이 최대공약수

Example 2) 78696과 19332의 최대공약수

- 78696 / 19332 $\rightarrow$ 1368  
- 19332 / 1368 $\rightarrow$ 180  
- 1368 / 180 $\rightarrow$ 108  
- 180 / 108 $\rightarrow$ 72  
- 108 / 72 $\rightarrow$ 36  
- 72 / 36 $\rightarrow$ 0  $\rightarrow$ 36 이 최대공약수

두개의 정수 m 과 n 이 있을 경우 m % n = 0 이면 n 이 GCD (Greatest Common Divisor, 최대 공약수) 이다.

<table>
    <tr>
        <td>m</td>
        <td>n</td>   
    </tr>
    <tr>
        <td>44</td>
        <td>12</td>    
    </tr>
    <tr>
        <td>36</td>
        <td></td>    
    </tr>
    <tr>
        <td>8</td>
        <td>12</td>    
    </tr>
    <tr>
        <td></td>
        <td>8</td>    
    </tr>
    <tr>
        <td>8</td>
        <td>4</td>    
    </tr>
    <tr>
        <td>0</td>
        <td>4</td>
        <td>GCD</td>
    </tr>
</table>

- 원리 : m % n = 0 가 될 때까지 작은 수(나머지)로 큰 수를 계속 나누어준다. 나머지가 0 이 될 때 나누어준 수가 GCD 이다.

In [ ]:
def gcd(m, n):
    if m < n:
        print('큰수를 먼저 입력해 주세요')
        return
    while m % n != 0:
    
    # CODE HERE

gcd(44, 12)

<div>











</div>

&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>
&nbsp;<br>